### PR TITLE
kubeadm: add kubelet fstype preflight check

### LINKED ISSUE #89181: Pods with subPath are stuck on Terminating
**What happened**:

Pods with subPath are stuck on Terminating when deleting

**What you expected to happen**:

Pod succefully removed.

**How to reproduce it (as minimally and precisely as possible)**:

crate:

```yaml
---
apiVersion: v1
kind: Pod
metadata:
  name: test
spec:
  containers:
  - image: docker.io/library/alpine
    name: test
    command: ['sleep', 'infinity']
    volumeMounts:
    - mountPath: /blabla/sss
      name: keys
      subPath: somekey
  volumes:
  - name: keys
    secret:
      secretName: test
---
apiVersion: v1
kind: Secret
metadata:
  name: test
data:
  somekey: R29vZCBhdHRlbXBsdCwgYnV0IG5vdCE=
  someotherkey: QXJlIHlvdSBraWRkaW5nIG1lPw==
```

remove the pod:

```
kubectl delete pod test
```

check the kubelet logs:

```
Mar 17 20:40:31 m1c4 kubelet[13866]: E0317 20:40:31.230957   13866 nestedpendingoperations.go:270] Operation for "\"kubernetes.io/secret/89b3eaa9-0bd8-4cc6-89a0-2fd9a346e238-keys\" (\"89b3eaa9-0bd8-4cc6-89a0-2fd9a346e238\")" failed. No retries permitted until 2020-03-17 20:42:33.230929328 +0100 CET m=+1192722.291304745 (durationBeforeRetry 2m2s). Error: "error cleaning subPath mounts for volume \"keys\" (UniqueName: \"kubernetes.io/secret/89b3eaa9-0bd8-4cc6-89a0-2fd9a346e238-keys\") pod \"89b3eaa9-0bd8-4cc6-89a0-2fd9a346e238\" (UID: \"89b3eaa9-0bd8-4cc6-89a0-2fd9a346e238\") : error processing /var/lib/kubelet/pods/89b3eaa9-0bd8-4cc6-89a0-2fd9a346e238/volume-subpaths/keys/test: error cleaning subpath mount /var/lib/kubelet/pods/89b3eaa9-0bd8-4cc6-89a0-2fd9a346e238/volume-subpaths/keys/test/0: unmount failed: exit status 32\nUnmounting arguments: /var/lib/kubelet/pods/89b3eaa9-0bd8-4cc6-89a0-2fd9a346e238/volume-subpaths/keys/test/0\nOutput: umount: /var/lib/kubelet/pods/89b3eaa9-0bd8-4cc6-89a0-2fd9a346e238/volume-subpaths/keys/test/0: not mounted.\n"
Mar 17 20:42:33 m1c4 kubelet[13866]: I0317 20:42:33.234534   13866 reconciler.go:183] operationExecutor.UnmountVolume started for volume "keys" (UniqueName: "kubernetes.io/secret/89b3eaa9-0bd8-4cc6-89a0-2fd9a346e238-keys") pod "89b3eaa9-0bd8-4cc6-89a0-2fd9a346e238" (UID: "89b3eaa9-0bd8-4cc6-89a0-2fd9a346e238")
Mar 17 20:42:33 m1c4 kubelet[13866]: E0317 20:42:33.257819   13866 nestedpendingoperations.go:270] Operation for "\"kubernetes.io/secret/89b3eaa9-0bd8-4cc6-89a0-2fd9a346e238-keys\" (\"89b3eaa9-0bd8-4cc6-89a0-2fd9a346e238\")" failed. No retries permitted until 2020-03-17 20:44:35.257684981 +0100 CET m=+1192844.318060559 (durationBeforeRetry 2m2s). Error: "error cleaning subPath mounts for volume \"keys\" (UniqueName: \"kubernetes.io/secret/89b3eaa9-0bd8-4cc6-89a0-2fd9a346e238-keys\") pod \"89b3eaa9-0bd8-4cc6-89a0-2fd9a346e238\" (UID: \"89b3eaa9-0bd8-4cc6-89a0-2fd9a346e238\") : error processing /var/lib/kubelet/pods/89b3eaa9-0bd8-4cc6-89a0-2fd9a346e238/volume-subpaths/keys/test: error cleaning subpath mount /var/lib/kubelet/pods/89b3eaa9-0bd8-4cc6-89a0-2fd9a346e238/volume-subpaths/keys/test/0: unmount failed: exit status 32\nUnmounting arguments: /var/lib/kubelet/pods/89b3eaa9-0bd8-4cc6-89a0-2fd9a346e238/volume-subpaths/keys/test/0\nOutput: umount: /var/lib/kubelet/pods/89b3eaa9-0bd8-4cc6-89a0-2fd9a346e238/volume-subpaths/keys/test/0: not mounted.\n"
```

**Anything else we need to know?**:

**Environment**:
- Kubernetes version (use `kubectl version`):
```
Client Version: version.Info{Major:"1", Minor:"17", GitVersion:"v1.17.0", GitCommit:"70132b0f130acc0bed193d9ba59dd186f0e634cf", GitTreeState:"clean", BuildDate:"2019-12-07T21:20:10Z", GoVersion:"go1.13.4", Compiler:"gc", Platform:"linux/amd64"}
Server Version: version.Info{Major:"1", Minor:"17", GitVersion:"v1.17.0", GitCommit:"70132b0f130acc0bed193d9ba59dd186f0e634cf", GitTreeState:"clean", BuildDate:"2019-12-07T21:12:17Z", GoVersion:"go1.13.4", Compiler:"gc", Platform:"linux/amd64"}
```
- Cloud provider or hardware configuration: Bare Metal
- OS (e.g: `cat /etc/os-release`):
```
NAME="Ubuntu"
VERSION="18.04.3 LTS (Bionic Beaver)"
ID=ubuntu
ID_LIKE=debian
PRETTY_NAME="Ubuntu 18.04.3 LTS"
VERSION_ID="18.04"
HOME_URL="https://www.ubuntu.com/"
SUPPORT_URL="https://help.ubuntu.com/"
BUG_REPORT_URL="https://bugs.launchpad.net/ubuntu/"
PRIVACY_POLICY_URL="https://www.ubuntu.com/legal/terms-and-policies/privacy-policy"
VERSION_CODENAME=bionic
UBUNTU_CODENAME=bionic
```
- Kernel (e.g. `uname -a`):
```
Linux m1c4 4.15.18-24-pve #1 SMP PVE 4.15.18-52 (Thu, 05 Dec 2019 10:14:17 +0100) x86_64 x86_64 x86_64 GNU/Linux
```
- Install tools: kubeadm
- Network plugin and version (if this is a network-related bug):
- Others:
docker version: 19.3.5

docker is running with the next config:
```json
{
  "exec-opts": ["native.cgroupdriver=systemd"],
  "iptables": false,
  "ip-forward": false,
  "bridge": "none",
  "log-driver": "journald",
  "storage-driver": "overlay2"
}
```

kubelet versions:
```
Kubernetes v1.17.1
Kubernetes v1.17.4
```

created file is empty and not mountpoint, path is not symlink:
```bash
# mountpoint /var/lib/kubelet/pods/89b3eaa9-0bd8-4cc6-89a0-2fd9a346e238/volume-subpaths/keys/test/0 
/var/lib/kubelet/pods/89b3eaa9-0bd8-4cc6-89a0-2fd9a346e238/volume-subpaths/keys/test/0 is not a mountpoint

# ls -lah
-rw-r----- 1 root root 0 Mar 17 20:37 /var/lib/kubelet/pods/89b3eaa9-0bd8-4cc6-89a0-2fd9a346e238/volume-subpaths/keys/test/0

# readlink -f /var/lib/kubelet/pods/89b3eaa9-0bd8-4cc6-89a0-2fd9a346e238/volume-subpaths/keys/test/0 
/var/lib/kubelet/pods/89b3eaa9-0bd8-4cc6-89a0-2fd9a346e238/volume-subpaths/keys/test/0
```

Not sure, but maybe problem caused by overlayfs root:

```bash
# df -h /var/lib/kubelet/pods/
Filesystem               Size  Used Avail Use% Mounted on
/run/initramfs/ltsp/cow   16G  412M   16G   3% /
```

### PR DESCRIPTION
**Pull request by kvaps**
**What type of PR is this?**

/kind feature

**What this PR does / why we need it**:

Kubelet is working incorrect on overlayfs, better to inform users in advance in preflight check.
This is an alternative to https://github.com/kubernetes/utils/pull/161

See https://kubernetes.slack.com/archives/C09QZFCE5/p1589544579387600 for more details.

**Which issue(s) this PR fixes**:

Fixes https://github.com/kubernetes/kubernetes/issues/89181#issuecomment-617326253 by dropping support of overlayfs

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:

```release-note
kubeadm: add a preflight check to error out if overlayfs is detected
```

**Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.**:

```docs

```

### CODE REVIEW

**Comment by kvaps** [comment]
/sig cluster-lifecycle
/area kubeadm
/assign @neolit123 
cc @msau42 @jsafrane

File: cmd/kubeadm/app/preflight/checks.go [hunk 1]
```diff
@@ -674,6 +676,31 @@ type etcdVersionResponse struct {
...
+// KubeletFSTypeCheck warns if wrong filesystem mounted on kubelet run directly
+type KubeletFSTypeCheck struct{}
+
+// Name will return "KubeletFSType" as name for KubeletFSTypeCheck
+func (KubeletFSTypeCheck) Name() string {
+	return "KubeletFSType"
+}
+
+// Check validates that kubelet is not running on a unsupported filesystem
+func (kfc KubeletFSTypeCheck) Check() (warnings, errorList []error) {
```
Comment (neolit123) [inline cmd/kubeadm/app/preflight/checks.go:688]: i'm pretty sure that this has to go in _linux.go and be a NO-OP in _windows.go.

**Comment by neolit123** [comment]
unless a kubelet maintainer / sig-storage confirm that "we have no plans to support  overlayfs" i don't think we should add this check to gate on the side of kubeadm. this seems like a bug (?) that maybe can be resolved elsewhere and the purpose of preflight checks is not to gate against bugs.

> Dropping support of OverlayFS

this release note should be "kubeadm: add a preflight check to error out if overlayfs is detected"

**Comment by kvaps** [comment]
> unless a kubelet maintainer / sig-storage confirm that "we have no plans to support overlayfs" i don't think we should add this check to gate on the side of kubeadm. this seems like a bug (?) that maybe can be resolved elsewhere and the purpose of preflight checks is not to gate against bugs.

Sure, I agree.
The solution for this bug is not using `IsLikelyNotMountPoint` check for the files located on OverlayFS but use more expensive `IsNotMountPoint()` check, which might lead performance issues on another types of filesystems.
The reason for this check is to inform people about possible problems of running kubelet on top of OverlayFS.

> this release note should be "kubeadm: add a preflight check to error out if overlayfs is detected"

fixed

**Comment by rosti** [comment]
@neolit123 The original overlayfs storage driver seems to be deprecated in Docker 18.09:

```
```

**Comment by neolit123** [comment]
@kvaps so this means the preflight check should only happen in the case of dockershim?

**Comment by kvaps** [comment]
No, this is unrelated with docker or any other CRI.

This preflight is needed to make sure that `/var/lib/kubelet` is not mounted in ovelayfs.

**Comment by neolit123** [comment]
> This preflight is needed to make sure that /var/lib/kubelet is not mounted in ovelayfs, 

ok, in that case the error message in the check should indicate this is about /var/lib/kubelet .

> because implemented in Kubelet IsLikelyNotMountPoint() check is not working correct with ovelayfs.

still feels like a bug that we can resolve instead of adding this potentially temporary preflight check.

**Comment by kvaps** [comment]
> ok, in that case the error message in the check should indicate this is about /var/lib/kubelet .

it is 
https://github.com/kubernetes/kubernetes/blob/ad031e6b68526a655d0015b9197c204f52c21253/cmd/kubeadm/app/preflight/checks.go#L697

> still feels like a bug that we can resolve instead of adding this potentially temporary preflight check.

I agree, here is purposed solution https://github.com/kubernetes/utils/pull/161

**Comment by neolit123** [comment]
> I agree, here is purposed solution kubernetes/utils#161

looks like that PR had gone stale at some point.
you could remind the #sig-storage channel about it or join their Zoom meeting:
https://github.com/kubernetes/community/tree/master/sig-storage

**Comment by neolit123** [comment]
should be resolved in k/utils.
/close

**Pull request closed without merging**